### PR TITLE
Fix Dimensions not updating on Android

### PR DESCRIPTION
**Pull request by jonnyandrew**
## Summary

When retrieving the device dimensions through the JS `Dimensions` utility, the result of `Dimensions.get` can be incorrect on Android.

### Related issues

- https://github.com/facebook/react-native/issues/29105
- https://github.com/facebook/react-native/issues/29451
- https://github.com/facebook/react-native/issues/29323

The issue is caused by the Android `DeviceInfoModule` that provides initial screen dimensions and then subsequently updates those by emitting `didUpdateDimensions` events. The assumption in that implementation is that the initial display metrics will not have changed prior to the first check for updated metrics. However that is not the case as the device may be rotated (as shown in the attached video).

The solution in this PR is to keep track of the initial dimensions for comparison at the first check for updated metrics.

## Changelog

[Android] [Fixed] - Fix Dimensions not updating

### Steps to reproduce
1. Install the RNTester app on Android from the `main` branch.
2. Set the device auto-rotation to ON
3. Start the RNTester app
4. While the app is loading, rotate the device
5. Navigate to the `Dimensions` screen
6. Either
 a. Observe the screen width and height are reversed, or
 b. Quit the app and return to step 3.

#### Manually
Using the above steps, the issue should no longer be reproducible.

#### Automatically
See unit tests in `ReactAndroid/src/test/java/com/facebook/react/modules/deviceinfo/DeviceInfoModuleTest.java`

### Video

https://user-images.githubusercontent.com/4940864/128485453-2ae04724-4ac5-4267-a59a-140cc3af626b.mp4

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/modules/deviceinfo/DeviceInfoModule.java [hunk 1]
```diff
@@ -85,8 +90,7 @@ public void emitUpdateDimensionsEvent() {
 
     if (mReactApplicationContext.hasActiveReactInstance()) {
       // Don't emit an event to JS if the dimensions haven't changed
-      WritableNativeMap displayMetrics =
-          DisplayMetricsHolder.getDisplayMetricsNativeMap(mFontScale);
+      WritableMap displayMetrics = DisplayMetricsHolder.getDisplayMetricsWritableMap(mFontScale);
```
Comment (lunaleaps) [inline ReactAndroid/src/main/java/com/facebook/react/modules/deviceinfo/DeviceInfoModule.java:93]: What was the logic behind `WritableNativeMap` and `WritableMap`? Sorry I'm not familiar with the difference

**Review by lunaleaps — CHANGES_REQUESTED** [review]
Interesting, it seems we have a lot of unused code in DisplayMetrics. I'm wondering if you've tested as well with split screen and verified there are no changes with that behavior.

File: ReactAndroid/src/main/java/com/facebook/react/modules/deviceinfo/DeviceInfoModule.java (hunk 1 above)
Comment (jonnyandrew) [reply ReactAndroid/src/main/java/com/facebook/react/modules/deviceinfo/DeviceInfoModule.java:93]: `WritableNativeMap` is an implementation of `WritableMap` (and that is the implementation that is returned from `DisplayMetricsHolder`). But in Java-only test code, `WritableNativeMap` is not supported and needs to be switched out for `JavaOnlyMap`. So in order to test this class, it needs to depend on `WritableMap` instead.

**Comment by jonnyandrew** [comment]
Thanks for the review @lunaleaps! I have not tested in split screen but I will take a look. And if you have any particular split screen test cases in mind, please let me know.

**Comment by jonnyandrew** [comment]
@lunaleaps I have run a few manual tests in split screen mode as requested, and I did not see any change in behaviour (except this fix as expected).

| Test case | Before | After |
|-----------|--------|-------|
| 1. Given app is not in split screen, when entering split screen, then dimensions are not changed | Pass ✅ | Pass ✅ |
| 2. Given app is in split screen, when rotated, then dimensions are reversed | Pass ✅ | Pass ✅ |
| 3. Given app is initialising in split screen, when rotated, then final dimensions are correct | Flaky ❌ | Pass ✅ |

I ran the tests on a Pixel 3a emulator with Android 10. Let me know if you'd like to post the video recordings.

**Pull request closed without merging**